include/dlaf/matrix/internal/tile_future_manager.h

```diff
@@ -0,0 +1,85 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include <hpx/hpx.hpp>
+#include "dlaf/matrix/tile.h"
+
+namespace dlaf {
+namespace matrix {
+namespace internal {
+
+template <class ReturnTileType, class TileType>
+hpx::future<ReturnTileType> setPromiseTileFuture(hpx::future<TileType>& old_future,
+                                                 hpx::lcos::local::promise<TileType>& p) noexcept {
+  return old_future.then(hpx::launch::sync, [p = std::move(p)](hpx::future<TileType>&& fut) mutable {
+    std::exception_ptr current_exception_ptr;
+
+    try {
+      return ReturnTileType(std::move(fut.get().setPromise(std::move(p))));
+    }
+    catch (...) {
+      current_exception_ptr = std::current_exception();
+    }
+
+    // The exception is set outside the catch block since set_exception may
+    // yield. Ending the catch block on a different worker thread than where it
+    // was started may lead to segfaults.
+    p.set_exception(current_exception_ptr);
+    std::rethrow_exception(current_exception_ptr);
+  });
+}
+
+template <class ReturnTileType, class TileType>
+hpx::future<ReturnTileType> getTileFuture(hpx::future<TileType>& tile_future) noexcept {
+  hpx::future<TileType> old_future = std::move(tile_future);
+  hpx::lcos::local::promise<TileType> p;
+  tile_future = p.get_future();
+  return setPromiseTileFuture<ReturnTileType>(old_future, p);
+}
+
+template <class T, Device device>
+class TileFutureManager {
+public:
+  using TileType = Tile<T, device>;
+  using ConstTileType = Tile<const T, device>;
+
+  TileFutureManager() {}
+
+  TileFutureManager(TileType&& tile) : tile_future_(hpx::make_ready_future(std::move(tile))) {}
+
+  hpx::shared_future<ConstTileType> getReadTileSharedFuture() noexcept {
+    if (!tile_shared_future_.valid()) {
+      tile_shared_future_ = std::move(getTileFuture<ConstTileType>(tile_future_));
+    }
+    return tile_shared_future_;
+  }
+
+  hpx::future<TileType> getRWTileFuture() noexcept {
+    tile_shared_future_ = {};
+    return getTileFuture<TileType>(tile_future_);
+  }
+
+  void clearSync() {
+    tile_shared_future_ = {};
+    tile_future_.get();
+  }
```

Comment:
The name of this function is 100% clear to me. Is it guaranteed that the non-shared future is always the active one here? Is an equivalent for the shared future required? Would an assert be appropriate if there are indeed preconditions?

Reply:
I suppose that you mean "not 100% clear to me"...

The function clear the content of the object in a synchronous way.
This is needed to guarantee that the matrix object is not destroyed (and therefore deallocates memory) before the work on the tiles is completed.

By design, the only reason why tile_future_ wouldn't be active here is that clearSync has been called before. However this applies to all methods of this object.

I will add some doc about this.

Reply:
Yes, that's indeed what I meant, sorry... I'll rephrase: is it not possible that `tile_shared_future_` is the active future when the `Matrix` destructor is called, not `tile_future_`? I.e. if the last access to a tile is read-only, `tile_future_` is moved into `tile_shared_future_`, and `get` on the now invalid `tile_future_` would fail. What am I missing? Note: I'm only asking because I may simply not have the full picture, and this is in fact impossible due to other constraints.

Reply:
Please note that `tile_future_` is always active and valid, even when a shared_future is available. 

Please see [misc/synchronization.md](https://github.com/eth-cscs/DLA-Future/blob/master/misc/synchronization.md#matrix-mechanism) especially figures 1 and 2 to understand how the mechanism works.

Reply:
Ah, I see now. Thanks! In that case the comment about the variant indeed doesn't make sense.